netty/src/main/java/io/grpc/netty/GrpcHttp2Headers.java

```diff
@@ -0,0 +1,146 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.netty;
+
+import io.netty.handler.codec.http2.Http2Headers;
+import io.netty.util.AsciiString;
+
+import java.util.Iterator;
+import java.util.Map.Entry;
+import java.util.NoSuchElementException;
+
+/**
+ * A custom implementation of Http2Headers that only includes methods used by gRPC.
+ */
+final class GrpcHttp2Headers extends AbstractHttp2Headers {
+
+  private final AsciiString[] normalHeaders;
+  private final AsciiString[] preHeaders;
+  private static final AsciiString[] EMPTY = new AsciiString[]{};
+
+  static GrpcHttp2Headers clientHeaders(byte[][] serializedMetadata, AsciiString authority,
+      AsciiString path, AsciiString method, AsciiString scheme, AsciiString userAgent) {
+    AsciiString[] preHeaders = new AsciiString[] {
+        Http2Headers.PseudoHeaderName.AUTHORITY.value(), authority,
+        Http2Headers.PseudoHeaderName.PATH.value(), path,
+        Http2Headers.PseudoHeaderName.METHOD.value(), method,
+        Http2Headers.PseudoHeaderName.SCHEME.value(), scheme,
+        Utils.CONTENT_TYPE_HEADER, Utils.CONTENT_TYPE_GRPC,
+        Utils.TE_HEADER, Utils.TE_TRAILERS,
+        Utils.USER_AGENT, userAgent,
+    };
+    return new GrpcHttp2Headers(preHeaders, serializedMetadata);
+  }
+
+  static GrpcHttp2Headers serverHeaders(byte[][] serializedMetadata) {
+    AsciiString[] preoHeaders = new AsciiString[] {
+        Http2Headers.PseudoHeaderName.STATUS.value(), Utils.STATUS_OK,
+        Utils.CONTENT_TYPE_HEADER, Utils.CONTENT_TYPE_GRPC,
+    };
+    return new GrpcHttp2Headers(preoHeaders, serializedMetadata);
+  }
+
+  static GrpcHttp2Headers serverTrailers(byte[][] serializedMetadata) {
+    return new GrpcHttp2Headers(EMPTY, serializedMetadata);
+  }
+
+  private GrpcHttp2Headers(AsciiString[] preHeaders, byte[][] serializedMetadata) {
+    normalHeaders = new AsciiString[serializedMetadata.length];
+    for (int i = 0; i < normalHeaders.length; i++) {
+      normalHeaders[i] = new AsciiString(serializedMetadata[i], false);
+    }
+    this.preHeaders = preHeaders;
+  }
+
+  @Override
+  public Iterator<Entry<CharSequence, CharSequence>> iterator() {
+    return new Itr();
+  }
+
+  @Override
+  public int size() {
+    return (normalHeaders.length + preHeaders.length) / 2;
+  }
+
+  private class Itr implements Iterator<Entry<CharSequence, CharSequence>> {
```

Comment:
To further optimize the headers,  you could have the `Itr` class additionally extend from `Entry<CharSequence, CharSequence>` and on each next call always return `this` with updated values. That way iterating through the headers (as I believe we do when writing the headers on the wire) only creates a single object, instead of an `Iterator` + `n` `Entry` objects.